### DISCUSSION TITLE
Special-case functools.wraps decorator support

### DISCUSSION THREAD

**Discussion by ARF1** [microsoft/pylance-release#2716] [Enhancement]
Currently a function decorated with `@functools.wraps(...)` wreaks havoc with intellisense. This is particularly unfortunate, since `@wraps` specifically exists in python to support proper docstrings with wrapped functions.

According to the [python documentation](https://docs.python.org/2/library/functools.html#functools.update_wrapper) `@wraps`  does (by default) the following:

- assigns to the wrapper function’s `__name__`, `__module__` and `__doc__` to the wrapped function`s values
- updates the wrapper function’s `__dict__` with the wrapped function's `__dict__`

Intellisense should also be able to handle the frequently used python idiom with `*args, **kwargs` in the wrapper:

```
from functools import wraps

def foo(param1: str, param2: int = 1) -> float:
    "Some docstring"
    pass

@wraps(foo)
def bar(*arg, **kwargs):
    print('calling foo')
    return foo(*args, **kwargs)
```

Intellisense type hints that show `bar: _AnyCallable` and parameter hints that show `(*args, **kwargs) -> Any`, as is currently the case, are completely pointless.

I realize that special-casing is not a very attractive prospect, but at least for this standard-library idiom specifically designed for making function wrapping more transparent, Intellisense really should work out of the box.

**Comment by Strauman** [comment]
This issue is similar and is transferable to e.g. `functools.partial`

**Comment by bschnurr** [comment]
If you specify parameters on bar in your example the signature appears to work.

![image](https://user-images.githubusercontent.com/1946977/117330158-41390180-ae4a-11eb-8c73-fa3e4da7297f.png)

forgot the return annotation.

**Comment by bschnurr** [comment]
> This issue is similar and is transferable to e.g. `functools.partial`

can you create a new issue with an example and link it?

**Comment by ARF1** [comment]
@bschnurr 

> If you specify parameters on bar in your example the signature appears to work.

That has always been the case but it defeats the point of `wraps` imho: If you then also duplicate the docstring  you can avoid using `wraps` altogether. ;-)

**Comment by bschnurr** [comment]
this might be an alternative, use `update_wrapper` directly 
![image](https://user-images.githubusercontent.com/1946977/117485560-5763c200-af1d-11eb-9053-ad664ca64bfd.png)

**Comment by bschnurr** [comment]
making this change to functools.pyi appears to do what you want, and doc strings show up!

```
def wraps(wrapped: _AnyCallable, assigned: Sequence[str] = ..., updated: Sequence[str] = ...) -> Callable[[_T], _T]: ...

To

def wraps(wrapped: _T, assigned: Sequence[str] = ..., updated: Sequence[str] = ...) -> Callable[[_T], _T]: ...
```

looks like we need to update typeshed

![image](https://user-images.githubusercontent.com/1946977/117488715-972ca880-af21-11eb-99bb-15177cb8bdb7.png)

**Comment by jakebailey** [comment]
That does change the behavior of the stub, though, and I don't think it's equivalent.

**Comment by bschnurr** [comment]
Ya. I think you're right. 

@erictraut any thoughts?

**Comment by bschnurr** [comment]
I tried requesting a change to typeshed but they pointed out that the signature defaults don't actually get updated.

https://github.com/python/typeshed/pull/5403#issuecomment-837050443
https://bugs.python.org/issue41232

Maybe you can define your own wraps function?

**Comment by bschnurr** [comment]
maybe there is something we can do to support docstrings from updating though. looking at `__wrapped__`

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by liangli2718** [comment]
I am bothered by the same issue. A side remark is that Pylance v2023.3.10 has the correct behavior, higher versions not.

**Comment by Llamrei** [comment]
Just bumping that this would still be a nice feature, especially for deprecation/function renaming refactors - enabling maintenance of docstrings in one place:
```python
def new_f(a: int):
    """How this works"""

@wraps(new_f)    
@deprecated
def old_name(a: int):
    return new_f(a)
```
And a hover of `old_name` would give the docs of `new_f`. Frankly the ideal would be:

```python
def new_f(a: int, b:int, c:float):
    """How this works"""

@wraps(new_f)    
@deprecated
def old_name(*args, **kwargs):
    return new_f(*args, **kwargs)
```
